feat(plugin): Refactor NPM plugin (#719)
canonical/craft-parts: docs/common/craft-parts/reference/plugins/npm_plugin.rst

.. _craft_parts_npm_plugin:

> NPM Plugin

Reviewer: This is a great doc page, thanks a lot! Can you add a quick example at the end, showcasing some of the properties? 
some reference: https://github.com/canonical/craft-parts/blob/new-plugin-docs/docs/common/craft-parts/reference/plugins/go_plugin.rst?plain=1#L73

=============

The NPM plugin can be used for Node.js projects that use NPM (or Yarn) as the package manager.

Keywords
--------

In addition to the common :ref:`plugin <part-properties-plugin>` and
:ref:`sources <part-properties-sources>` keywords, this plugin provides the following
plugin-specific keywords:

npm-include-node
~~~~~~~~~~~~~~~~~~
**Type:** boolean
**Default:** False

If this option is set to ``true``, the plugin will download and include the 

> Node.js binaries and its dependencies in the Snap.

Reviewer 2: ```suggestion
Node.js binaries and its dependencies in the resulting package.
```

Since craft-parts doesn't only build snaps this is application neutral.

If ``npm-include-node`` is true, then :ref:`npm-node-version` must be defined.

.. _npm-node-version:

npm-node-version
~~~~~~~~~~~~~~~~~~~
**Type:** string
**Default:** ``null``

> Which version of Node.js to download and include in the final Snap.

Reviewer 2: ```suggestion
Which version of Node.js to download and include in the final package.
```

Required if ``npm-include-node`` is set to true.

> The option accepts a NVM-style version string, you can specify one of:

Reviewer 2: ```suggestion
The option accepts an NVM-style version string; you can specify one of:
```

.. _craft_parts_npm_plugin:

NPM Plugin
=============

The NPM plugin can be used for Node.js projects that use NPM (or Yarn) as the package manager.

Keywords
--------

In addition to the common :ref:`plugin <part-properties-plugin>` and
:ref:`sources <part-properties-sources>` keywords, this plugin provides the following
plugin-specific keywords:

npm-include-node
~~~~~~~~~~~~~~~~~~
**Type:** boolean
**Default:** False

If this option is set to ``true``, the plugin will download and include the 
Node.js binaries and its dependencies in the resulting package.
If ``npm-include-node`` is true, then :ref:`npm-node-version` must be defined.

.. _npm-node-version:

npm-node-version
~~~~~~~~~~~~~~~~~~~
**Type:** string
**Default:** ``null``

Which version of Node.js to download and include in the final package.
Required if ``npm-include-node`` is set to true.

The option accepts an NVM-style version string; you can specify one of:

> * exact version (e.g. "20.12.2")

Reviewer 3: ```suggestion
* exact version (e.g. ``"20.12.2"``)
```

> * major+minor version (e.g. "20.12")

Reviewer 3: ```suggestion
* major+minor version (e.g. ``"20.12"``)
```

> * major version (e.g. "20")

Reviewer 3: ```suggestion
* major version (e.g. ``"20"``)
```

> * LTS code name (e.g. "lts/iron")

Reviewer 3: ```suggestion
* LTS code name (e.g. ``"lts/iron"``)
```

> * latest mainline version ("node")

Reviewer 3: ```suggestion
* latest mainline version (``"node"``)
```

.. warning::
    In the ``nvm`` utility, you can specify "system" to use system
    Node.js package, but this is unsupported in this plugin, as we
    are using upstream Node.js binaries.

    Also, the ``iojs`` specifier is unsupported in this plugin,
    as the ``iojs`` project was merged back to Node.js circa. 2015.
    Using very old ``iojs`` runtime poses a significant security
    hazard. If your project still requires a JavaScript runtime
    from nearly a decade ago, you should seriously considering
    migrating to the modern Node.js runtime.

Examples
--------

The following example declares a part using the ``npm`` plugin.
In this example, we show how you may build the ``terser`` utility
(an utility for compressing and obfuscating JavaScript code).
It uses the latest mainline stable version of Node.js and includes
a copy of the Node.js runtime inside the final package.

.. code-block:: yaml

    parts:
        app:
            plugin: npm
            source: https://github.com/terser/terser
            source-type: git
            npm-include-node: true
            npm-node-version: "node"

Another example that shows how to install an application that

> published to npm registry but does not require Node.js runtime

Reviewer: ```suggestion
is published to the npm registry but does not require a Node.js runtime
```
